docs/RFCS/certificate_rotation.md

```diff
@@ -0,0 +1,293 @@
+
+# Online certificate rotation
+
+A good description of online key rotation in Go can be found in
+[Hitless TLS Certificate Rotation in Go](https://diogomonica.com/2017/01/11/hitless-tls-certificate-rotation-in-go/)
+
+## Adding a new CA certificate
+
+All TLS connections are given a [tls.Config](https://golang.org/pkg/crypto/tls/#Config) object
+containing a [*x509.CertPool](https://golang.org/pkg/crypto/x509/#CertPool) to verify
+server certificates (`tls.Config.RootCAs`) and client certificates (`tls.Config.ClientCAs`).
+
+Adding a new certificate to the pool can be done as long as we have access to the `tls.Config` or
+the `x509.CertPool`. Neither is safe for concurrent use so we will need a small wrapper to allow
+safe modification of the pool while serving connections.
+
+## Rotating node certificate
+
+Node certificates are tricker than CA certificates as we must pick exactly one to present.
+
+The `tls.Config` object allows us to specify node certificates in two ways:
+* set the certificate in `tls.Config.Certificates []Certificate`
+* set a per-connection callback in `tls.Config.GetCertificate func(*ClientHelloInfo) (*Certificate, error)`
+
+Modifying either will change which certificate is presented at session initiation or renegotiation.
+The `Certificates` object is not safe for concurrent use, making `GetCertificate` preferable.
+
+# Additional interfaces
+
+## Command line
+
+We will need to modify all commands that use certs to make use of the new directory structure.
+
+We will also need:
+* modification to `cert create-ca` to use an existing key.
+* `cert list` to list all CA certs and node/client cert/key pairs.
+
+## Admin UI / debug pages
+
+We want at least barebones listing of all certs on a given node, including validity
+dates, certificate chain (corresponding CA for a node cert), and valid hosts.
+
+Soon-to-expire certificates (or chains) must be reported prominently on the admin UI and
+available through external metrics.
+
+# Unresolved questions
+
+## `tls.Config` fields
+
+We need to verify that the proposed CA and node cert rotation mechanisms work, especially through
+grpc.
+We may be able to modify the `tls.Config.Certificates` field, but this would require a wrapper
+to allow concurrent use.
+On the other hand, using the `tls.Config.GetCertificate` callback mechanism **seems** to work
+(tested superficially by mberhault on 2017/03/18).
+
+## Renegotiation and certificate rotation
+
+Renegotiation may cause new certificates to be presented. We need to make sure this will not
+cause issues.
+
+The `tls.Config` comments also mention this happening in TLS 1.3:
+```
+GetClientCertificate may be called multiple times for the same
+connection if renegotiation occurs or if TLS 1.3 is in use.
+```
+
+## Multi-certificate DER files and postgres clients
+
+The Go `lib/pq` will use all certificates found in `ca.crt`, but this may not be the
+case of other libraries.
+It may be safer to keep a single CA certificate per file.
```

Comment:
We can run an inventory of what's supported. the JDBC driver does the right thing I believe, and that would cover a lot of bases.